src/common/mailingListServices.jsx

```diff
@@ -0,0 +1,145 @@
+      />
+    )
+  if (success) return <Status type="info" message="You have been subscribed!" />
+  return null
+}
+
+/**
+ * Submit button that displays processing status
+ * @param {boolean} processing 
+ */
+const SubmitButton = ({ processing }) => (
+  <button className="submitButton" type="submit" disabled={processing}>
+    {processing ? "Processing..." : "Subscribe"}
+  </button>
+)
+
+/**
+ * Logic for subscribing an email address to a GovDelivery topic,
+ * leveraging consumerfinance.gov's /subscriptions/new endpoint.
+ * @param {string} endpoint API endpoint
+ * @param {string} topicId GovDelivery Topic ID
+ * @returns {Object} Page elements, setters, and event handlers
+ */
+export const useSubscriptionLogic = ({ endpoint, topicId }) => {
+  const [emailAddress, setEmailAddress] = useState("")
+  const [error, setError] = useState(null)
+  const [success, setSuccess] = useState(null)
+  const [invalid, setInvalid] = useState(false)
+  const [processing, setProcessing] = useState(false)
+
+  /**
+   * Submit a topic subscription request to the subscription endpoint
+   * @param {string} url API endpoint
+   * @param {string} content URL encoded request body
+   */
+  const requestSubscription = (url, content) => {
+    const request = new XMLHttpRequest()
+
+    request.onreadystatechange = function () {
+      if (request.readyState == XMLHttpRequest.DONE) {
+        if (request.status == 200) setSuccess(request.responseText)
+        else if (request.status == 400) setError("Error (400): Bad Request")
+        else setError(`Error (${request.status}): ${request.responseText}`)
+
+        setTimeout(() => setProcessing(false), 250)
+      }
+    }
+
+    request.open("POST", url)
+    request.setRequestHeader(
+      "Content-Type",
+      "application/x-www-form-urlencoded"
+    )
+    request.setRequestHeader("X-Requested-With", "XMLHttpRequest") // Required for cf.gov to send JSON response
+    request.send(content)
+  }
+
+  /** Handle user input */
+  const onEmailChange = (e) => {
+    setError(null)
+    setSuccess(null)
+    setInvalid(null)
+    e && setEmailAddress(e.target.value)
+  }
+
+  /** Handle form submission */
+  const onSubmit = (event) => {
+    event.preventDefault()
+    if (!validateEmailFormat(emailAddress)) {
+      setInvalid("Invalid email address")
+      setError(null)
+      setSuccess(null)
+      return
+    } else {
+      const content = `email=${emailAddress}&code=${topicId}`
+      setSuccess(null)
+      setError(null)
+      setProcessing(true)
+      requestSubscription(endpoint, content)
+    }
+  }
+
+  const currentStatus = <StatusMessage {...{ invalid, error, success }} />
+
+  const submitButton = <SubmitButton {...{ processing }} />
+
+  return {
+    emailAddress,
+    setEmailAddress,
+    success,
+    setSuccess,
+    error,
+    setError,
+    processing,
+    setProcessing,
+    invalid,
+    setInvalid,
+    onEmailChange,
+    onSubmit,
+    currentStatus,
+    submitButton,
```

Comment:
having components exported as part of the custom hook feels a bit awkward; also, this is a lot of exports... aside from the emailAddress, should the component using this be able to set the invalid, processing, success, error states?

Reply:
Agreed, that was a lot of exports.  I've reduced the list to those elements that are actually imported by our Signup components. 
 
As for having components exported from a hook, it feels like a reasonable solution for this use case.  We've got a couple of Signup forms that need to display the same components in different layouts.  Those shared components rely on, and display, the current state stored in the `useSubscriptionLogic` hook.  Having those shared components be pre-composed means we don't need to duplicate that composition logic in two places.